tensorflow-xla-runtime.spec

```diff
@@ -21,11 +22,17 @@ CXXFLAGS="-fPIC %{arch_build_flags} ${CMS_EIGEN_CXX_FLAGS}"
 %endif
 
 pushd tensorflow/xla_aot_runtime_src
+  # the CMakelists.txt file is not intended to produce a shared library so create a static one first
   cmake . -DCMAKE_CXX_FLAGS="${CXXFLAGS}" -DCMAKE_CXX_STANDARD=%{cms_cxx_standard} -DBUILD_SHARED_LIBS=OFF
   make %{makeprocesses}
+
+  # build the shared library, linking missing symbols from abseil
+  gcc -shared -o libtf_xla_runtime.so -Wl,--whole-archive libtf_xla_runtime.a -Wl,--no-whole-archive \
+    -L${ABSEIL_CPP_ROOT}/lib -labsl_strings -labsl_str_format_internal
```

Comment:
do you understand why there are missing symbols? May be it is not picking up our abseil-cpp properly?

Reply:
@riga, I changed the cmake file (CMakeLists.txt) to have
```
add_library(tf_xla_runtime SHARED
  $<TARGET_OBJECTS:tf_xla_runtime_objects>
)
``` 
and then looking at the missing `abseil` symbols, I get
```
                 U _ZN4absl12lts_2022062319str_format_internal10FormatPackB5cxx11ENS1_21UntypedFormatSpecImplENS0_4SpanIKNS1_13FormatArgImplEEE
                 U _ZN4absl12lts_2022062319str_format_internal13FormatArgImpl8DispatchIiEEbNS2_4DataENS1_24FormatConversionSpecImplEPv
                 U _ZN4absl12lts_2022062319str_format_internal13FormatArgImpl8DispatchISt17basic_string_viewIcSt11char_traitsIcEEEEbNS2_4DataENS1_24FormatConversionSpecImplEPv
                 U _ZN4absl12lts_202206239StrAppendEPNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKNS0_8AlphaNumE
```

all of these are defined in the `absl_strings` library.  We can patch [a] `xla_aot_runtime_src/CMakeLists.txt ` and then it should generate the shared library (properly linked to  absl::strings)

[a]
```
--- xla_aot_runtime_src/CMakeLists.txt    2024-03-24 08:28:34.000000000 +0100
+++ xla_aot_runtime_src/CMakeLists.txt      2024-03-25 11:17:58.108587945 +0100
@@ -14,6 +14,8 @@
   -Wno-sign-compare
 )
 
-add_library(tf_xla_runtime STATIC
+find_package(absl REQUIRED)
+add_library(tf_xla_runtime SHARED
   $<TARGET_OBJECTS:tf_xla_runtime_objects>
 )
+target_link_libraries(tf_xla_runtime absl::strings)
```

Reply:
> do you understand why there are missing symbols? May be it is not picking up our abseil-cpp properly?

Only partially. Indeed it seems that it didn't find the abseil-cpp libraries which is now fixed. For the missing symbols of the tsl library I think the reason is that the `tensorflow_xla_runtime` sources (shipped with `py3-tensorflow`) only provide a subset of the implementations defined by headers.

![tsl](https://github.com/cms-sw/cmsdist/assets/1908734/68ab6e93-dcf4-42d1-8eb9-42248995afb0)

These symbols aren't needed anyway - unlike the absl ones - hence the `-Wl,--unresolved-symbols=ignore-in-shared-libs` in the toolfile. (I guess one could also fetch and compile the missing sources, though at the risk of inflating the runtime with unnecessary symbols.)

<hr/>

I added a patch in 7838448 and removed the initial static build step.